### PR TITLE
SelectOption: Add method `getLabel()` and change method `setLabel()`

### PR DESCRIPTION
**Pull request by sukhwinder33445**
This PR allows to easily get the label of the selected option. `getContect()` returns an array of `ValidHtml[]` and makes it harder to get the label as string.

# Usage:
`$form->getElement('user')->getOption($form->getValue('user'))->getLabel()`

### CODE REVIEW

**Comment by nilmerg** [comment]
Please rebase here, there's a new test covering this change.

**Review by lippserd — COMMENTED** [review]
What is the purpose of this PR, i.e. why do you have to change the label after construction? Also, there's `setContent()` and `getContent()` which could just be used on the element.

**Comment by sukhwinder33445** [comment]
I noticed that label can be null in the constructor. It makes no sense to have an option without label. Maybe this should be fixed in a separate PR.

**Comment by lippserd** [comment]
> What is the purpose of this PR, i.e. why do you have to change the label after construction? Also, there's `setContent()` and `getContent()` which could just be used on the element.

So you have updated the description. I suppose now it's just a matter of `getLabel()`. What do you want do with the label?

**Comment by lippserd** [comment]
> I noticed that label can be null in the constructor. It makes no sense to have an option without label. Maybe this should be fixed in a separate PR.

We'll talk about that later. option can be used in several places and here and there it applies to have no value. For a select option, of course, this makes no sense. However, value could be empty, which then defaults to the content. I will create an issue for all this later. First, I want to understand what you want to do with the label.

**Comment by lippserd** [comment]
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/option

**Comment by lippserd** [comment]
@sukhwinder33445 This can be closed, right?

**Review by lippserd — CHANGES_REQUESTED** [review]
In addition to the requested changes, `setLabel()` must NOT be removed, see [here for a usage example](https://github.com/Icinga/icingaweb2-module-vspheredb/blob/master/library/Vspheredb/Web/Form/InfluxDbConnectionForm.php#L178).

 I would like to see the following tests:

* `testRendering`
* `testRenderingAfterSetLabel`
* `testGetLabel`
* `testGetLabelAfterSetLabel`

File: src/FormElement/SelectOption.php [hunk 1]
```diff
@@ -19,27 +22,45 @@ class SelectOption extends BaseHtmlElement
...
+        $this->label = $label;
 
         $this->getAttributes()->registerAttributeCallback('value', [$this, 'getValue']);
     }
 
     /**
-     * @param $label
+     * Set the label of the option
+     *
+     * @param string
```
Comment (nilmerg) [inline src/FormElement/SelectOption.php:33]: `@param` requires also the name of the param

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Review by nilmerg — APPROVED** [review]

File: src/FormElement/SelectOption.php [hunk 2]
```diff
@@ -11,6 +11,9 @@ class SelectOption extends BaseHtmlElement
     /** @var mixed */
     protected $value;
 
+    /** @var string Label of the option */
```
Comment (lippserd) [inline src/FormElement/SelectOption.php:14]: `string|null`

File: src/FormElement/SelectOption.php [hunk 3]
```diff
@@ -19,27 +22,45 @@ class SelectOption extends BaseHtmlElement
...
 
     /**
-     * @param $label
+     * Set the label of the option
+     *
+     * @param string $label
+     *
      * @return $this
      */
     public function setLabel($label)
```
Comment (lippserd) [inline src/FormElement/SelectOption.php:37]: `string $label` or `string $label = null` if we want to support resetting the label.

Comment (lippserd) [reply src/FormElement/SelectOption.php:37]: Add return type.

**Review by lippserd — CHANGES_REQUESTED** [review]

**Comment by nilmerg** [comment]
As we're changing this so fundamentally, may I ask what the purpose of a `SelectOption` without a value is? Once constructed, it's not possible to change it, so if not passed to the constructor, the value is permanently not set.
I'd also question that the label may not be set. What's the purpose of a `SelectOption` without a label?

**Comment by lippserd** [comment]
> As we're changing this so fundamentally, may I ask what the purpose of a `SelectOption` without a value is? Once constructed, it's not possible to change it, so if not passed to the constructor, the value is permanently not set. I'd also question that the label may not be set. What's the purpose of a `SelectOption` without a label?

`null` is already used as value: https://github.com/Icinga/icingaweb2-module-vspheredb/blob/master/library/Vspheredb/Web/Form/InfluxDbConnectionForm.php#L60

Not having a label is questionable though.

File: src/FormElement/SelectOption.php [hunk 4]
```diff
@@ -8,38 +8,61 @@ class SelectOption extends BaseHtmlElement
...
+    /** @var string|null Label of the option */
+    protected $label;
+
     /**
      * SelectOption constructor.
      * @param string|null $value
      * @param string|null $label
      */
-    public function __construct($value = null, $label = null)
+    public function __construct(string $value = null, string $label = null)
```
Comment (nilmerg) [inline src/FormElement/SelectOption.php:22]: Then I'd recommend this: (and the consequences for the other parts that follow)

```suggestion
    public function __construct(?string $value, string $label)
```

File: src/FormElement/SelectOption.php [hunk 5]
```diff
@@ -8,38 +8,62 @@ class SelectOption extends BaseHtmlElement
 {
     protected $tag = 'option';
 
-    /** @var mixed */
+    /** @var string|null Value of the option */
```
Comment (nilmerg) [inline src/FormElement/SelectOption.php:11]: Please use `?string` here as well.

File: src/FormElement/SelectOption.php [hunk 6]
```diff
@@ -8,38 +8,62 @@ class SelectOption extends BaseHtmlElement
...
     protected $value;
 
+    /** @var string Label of the option */
+    protected $label;
+
     /**
      * SelectOption constructor.
+     *
      * @param string|null $value
```
Comment (nilmerg) [inline src/FormElement/SelectOption.php:20]: Please use `?string`

File: src/FormElement/SelectOption.php [hunk 7]
```diff
@@ -8,38 +8,62 @@ class SelectOption extends BaseHtmlElement
...
      * @return string
      */
-    public function getValue()
+    public function getLabel(): string
+    {
+        return $this->label;
+    }
+
+    /**
+     * Get the label of the option
```
Comment (nilmerg) [inline src/FormElement/SelectOption.php:56]: The label?

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Review by nilmerg — APPROVED** [review]

**Pull request merged**